src/sample.cpp

```diff
@@ -89,5 +89,44 @@ std::shared_ptr<SampleProvider> sampleProviderFromOptions(const TracerOptions& o
   return std::shared_ptr<SampleProvider>{new DiscardRateSampler(1.0 - options.sample_rate)};
 }
 
+RulesSampler::RulesSampler() : sampling_limiter_(getRealTime, 100, 100.0, 1) {}
+
+void RulesSampler::addRule(RuleFunc f) { sampling_rules_.push_back(f); }
+
+SampleResult RulesSampler::sample(const std::string& environment, const std::string& service,
+                                  const std::string& name, uint64_t trace_id) {
+  auto limit_result = sampling_limiter_.allow();
```

Comment:
we should be matching on the rule first, we only need to rate limit traces which would be sampled (p1).